components/dashboard/src/service/json-rpc-scm-client.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { PartialMessage } from "@bufbuild/protobuf";
+import { PromiseClient } from "@connectrpc/connect";
+import { SCMService } from "@gitpod/public-api/lib/gitpod/v1/scm_connect";
+import { converter } from "./public-api";
+import { getGitpodService } from "./service";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import {
+    GetSCMTokenRequest,
+    GuessTokenScopesRequest,
+    SearchRepositoriesRequest,
+    ListSuggestedRepositoriesRequest,
+    GetSCMTokenResponse,
+    ListSuggestedRepositoriesResponse,
+    SearchRepositoriesResponse,
+    GuessTokenScopesResponse,
+} from "@gitpod/public-api/lib/gitpod/v1/scm_pb";
+
+export class JsonRpcScmClient implements PromiseClient<typeof SCMService> {
+    async getSCMToken({ host }: PartialMessage<GetSCMTokenRequest>): Promise<GetSCMTokenResponse> {
+        if (!host) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "host is required");
+        }
+        const response = await getGitpodService().server.getToken({ host });
+        return new GetSCMTokenResponse({
+            token: response ? converter.toSCMToken(response) : undefined,
```

Comment:
Should not we throw NotFound if there's no response token from server

Reply:
The thing is, we're optimistically asking for tokens for all visible providers by their hosts. Turning this into 404s would result in a lot error responses, but everything would seems to be fine. This is due to the interaction pattern with the Dashboard, where we read all visible providers first, then try to merge with token scopes.

Reply:
More [discussion](https://gitpod.slack.com/archives/C0463A2KQUU/p1700578936506869?thread_ts=1700578189.913169&cid=C0463A2KQUU) on this.